File: 8-haunted-wasteland/app/Main.hs
{-# LANGUAGE ViewPatterns #-}
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
{-# OPTIONS_GHC -Wno-name-shadowing #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}

module Main (main) where

import           Data.Map (Map, fromList, keys, (!))

data Node = Node { left :: String, right :: String }
  deriving (Show)

type Name = String

readNode :: String -> (Name, Node)
readNode line = let (name, drop 4 -> rest) = splitAt 3 line
                    (l, drop 2 -> rest2) = splitAt 3 rest
                    r = take 3 rest2
                 in (name, Node l r)

readNodes :: [String] -> Map Name Node
readNodes = fromList . map readNode

type Directions = [Char]
steps :: Directions
      -> Name  -- start
      -> Map Name Node
      -> Int
steps = steps' 0
  where
    steps' :: Int -> Directions -> Name -> Map Name Node -> Int
    steps' acc _ "ZZZ" _ = acc
    steps' acc ('L':ds) curr nodes = let node = nodes ! curr
                                         next = left node
                                      in steps' (acc + 1) ds next nodes
    steps' acc ('R':ds) curr nodes = let node = nodes ! curr
                                         next = right node
                                      in steps' (acc + 1) ds next nodes

-- main :: IO ()
-- main = do input <- readFile "input.txt"
--           let directionsStr : _ : nodesStr = lines input
--           let nodes = readNodes nodesStr
--           let directions = cycle directionsStr
--           let result = steps directions "AAA" nodes
--           print result

-- part 2

-- brute force won't do

steps' :: Directions
       -> Map Name Node
       -> Name           -- start
       -> Int            -- steps taken
steps' = steps'' 0
  where steps'' :: Int -> Directions -> Map Name Node -> Name -> Int
        steps'' acc _ _ (last -> 'Z') = acc
        steps'' acc ('L':ds) nodes curr = let node = nodes ! curr
                                              next = left node
                                           in steps'' (acc + 1) ds nodes next
        steps'' acc ('R':ds) nodes curr = let node = nodes ! curr
                                              next = right node
                                           in steps'' (acc + 1) ds nodes next

main :: IO ()
main = do input <- readFile "input.txt"
          (directionsStr : _ : nodesStr) <- return $ lines input
          let nodes = readNodes nodesStr
          let directions = cycle directionsStr
          let starts = filter ((== 'A') . last) $ keys nodes
          let steps = map (steps' directions nodes) starts
          let result = foldr1 lcm steps  -- who would have thought?
          print result

File: 8-haunted-wasteland/main.py
# pyright: strict

from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Tuple
from itertools import cycle
from math import lcm


class Direction(Enum):
    LEFT = 1
    RIGHT = 2


# 'L' -> LEFT
# 'R' -> RIGHT
def parse_direction(char: str) -> Direction:
    match char:
        case 'L': return Direction.LEFT
        case _: return Direction.RIGHT


# 'LRLRL' -> [LEFT, RIGHT, LEFT, RIGHT, LEFT]
def parse_directions(line: str) -> List[Direction]:
    return list(map(parse_direction, line))


@dataclass
class Node:
    left: str
    right: str


# 'AAA = (BBB, CCC)' -> ('AAA', Node('BBB', 'CCC'))
def parse_node(line: str) -> Tuple[str, Node]:
    [name, rest] = line.split(' = ')
    [left, right] = rest[1:-1].split(', ')
    return (name, Node(left=left, right=right))


def parse_nodes(paragraph: str) -> Dict[str, Node]:
    lines = paragraph.splitlines()
    nodes = map(parse_node, lines)
    return dict(nodes)


def parse_input(input: str) -> Tuple[List[Direction], Dict[str, Node]]:
    [line, paragraph] = input.split('\n\n')
    return (parse_directions(line), parse_nodes(paragraph))


def steps_to_ZZZ(nodes: Dict[str, Node], curr: str, directions: List[Direction]) -> int:
    instructions = cycle(directions)
    step = 0

    # I HATE THIS
    while curr != 'ZZZ':
        direction = next(instructions)
        node = nodes[curr]
        curr = node.left if direction == Direction.LEFT else node.right
        step += 1

    return step
    

def main_():
    handle = open('input.txt', 'r')
    text = handle.read()
    handle.close()
    (directions, nodes) = parse_input(text)
    steps = steps_to_ZZZ(nodes, 'AAA', directions)
    print(steps)
    

# part 2

def steps_to_z(nodes: Dict[str, Node], curr: str, directions: List[Direction]) -> int:
    instructions = cycle(directions)
    step = 0

    while curr[-1] != 'Z':
        direction = next(instructions)
        node = nodes[curr]
        curr = node.left if direction == Direction.LEFT else node.right
        step += 1

    return step


def main():
    handle = open('input.txt', 'r')
    text = handle.read()
    handle.close()
    (directions, nodes) = parse_input(text)
    names = nodes.keys()
    a_nodes = filter(lambda name: name[-1] == 'A', names)
    stepss = map(lambda curr: steps_to_z(nodes, curr, directions), a_nodes)
    print(lcm(*stepss))



if __name__ == '__main__':
    main()
